Replace FontAwesome icons with native SVGs

4lm opened the pull request:
~~**Please note:** This PR is currently in Draft mode. I will address some minor styling issues and fix three occurrences (plus related tests) in the (www) layer where logic improperly intersects with the outer context (flask) layer. Reviews will be requested in the coming days.~~

~~**Edit:** Plus adding my documentation in this PR to the Sphinx documentation build.~~

**Edit:** Finally, this PR is ready for review 🥳

This PR solves the first task from #1035. In short, I removed the link tag usage of FontAwesome that had a high cost, adding over 600 KB of external JavaScript (and, on top, with great certainty, tracked our users) and replaced it with a custom icon template implementation.

A dedicated section in `docs/development_guide.rst` (starting at line 644) explains the implementation and its use. You can generate the documentation locally by running `make html` and then opening `/build/html/index.html` in your web browser for better readability.

This is my first PR for a long time in this project, and therefore, I'm aware of my lack of the grand historical context of all the thoughts and decisions that made this code base what it is today. Therefore, I'm ready not to pass on the first try. So, please don't be shy, dig deep, ask, criticize, and recommend improvements to shape the code in this PR into something we can all happily live with!

Plan: 74ce01fe-6e37-4207-a91e-9d68e2b2e910 (29x)

Amittai-Aviram reviewed (commented):
Looks great so far!  I have just inserted a few small comments and questions, mostly (but not all) about English language stuff.  Could you give us a demo at a meeting at some point?

4lm commented:
> Looks great so far! I have just inserted a few small comments and questions, mostly (but not all) about English language stuff. Could you give us a demo at a meeting at some point?

Great! Thanks for taking the time 👍 I will answer each point.

4lm commented on arbeitszeit_flask/__init__.py:
```diff
@@ -79,8 +80,9 @@ def create_app(config: Any = None, db: Any = None, template_folder: Any = None)
     initialize_babel(app)
 
-    # Setup template filter
+    # Setup template filters
```
Thanks for the info. I will change it to "Set up template filters".

4lm commented on arbeitszeit_flask/filters.py:
```diff
@@ -0,0 +1,100 @@
+
+    Usage (most cases):
+    -------------------
+    Assuming an SVG icon file named `example.html` exists in the `ICON_PATH`
```
The icon files can be considered jinja2 HTML template files, even though they only contain HTML (not-well formed) svg-tag elements at the moment, and no jinja-specific control flow or variables are used (yet). Also, the implementation might change if we replace Bulma CSS-specific things like the ' .icon` or `is-(small|medium|big)` classes wrapped (mostly) in span tags at the moment. Therefore, I would like to leave it as is for now.

4lm replied:
Maybe improving the description wording here could be an improvement.

Amittai-Aviram replied:
Maybe just to say that they are template files rather than pure SVG files?  Otherwise, whatever you choose is fine by me.

4lm commented on arbeitszeit_flask/filters.py:
```diff
@@ -0,0 +1,100 @@
+    Usage (most cases):
+    -------------------
+    Assuming an SVG icon file named `example.html` exists in the `ICON_PATH`
+    directory, you can use the as `icon` registered `icon_filter` function
```
OK, then this needs clarification. In Flask (the web framework), you can register filter functions, which can be used globally in any of the Flask (Jinja2) HTML template files. Therefore, in this PR, the filter function named `icon_filter` gets registered as a template filter named `icon` to transform a string from - for example - "user" (via) `{{ "user"|icon }}` into another string which is, in our case, an HTML snippet containing a svg-tag element derived from a file named "user.html" in the `arbeitszeit_flask/templates/icons` folder.

4lm replied:
In other words, to not have to write {{ "name"|icon_filter }} (hence the "icon_filter" part) in the template each time because, from the usage of this filter in the template, it's clear that it is a filter function. Therefore, the filter function "icon_filter" is registered as an "icon" and used as {{ "name"|icon }} in that template.

But why not call the original function "icon" in the first place? Because of context. If I import and use this function somewhere in the Python code, it would need to be clearer to the (human) reader that this Python is a template filter function. Therefore, I would name all filter functions in the Python code by convention `<name>_filter`.

Amittai-Aviram replied:
Thank you!  (So is a "filter function" actually what I would call a replacement function?)

4lm commented on arbeitszeit_flask/filters.py:
```diff
@@ -0,0 +1,100 @@
+
+    Usage (extend & overwrite):
+    ------------------
+    If you want to extend or overwrite SVG attributes do following.
```
👍

4lm commented on arbeitszeit_flask/filters.py:
```diff
@@ -0,0 +1,100 @@
+
+        if "<svg" not in svg_content:
+            return Markup(
+                f'<!-- Icon "{icon_name}" does not contain valid SVG content -->'
```
As discussed in the meeting, we will put this in an extra issue and form a task group around the logging/monitoring issue. Therefore I would leave it as is for now.

4lm commented on arbeitszeit_flask/templates/auth/signup_member.html:
Yes, many of those elements could be extracted into Jinja2 template macros. I suggest doing so in another PR because the purpose of this PR is to remove the usage of FontAwesome, and we are already 60+ files with changes deep. So, less is more here, I guess 😉

4lm commented on arbeitszeit_flask/templates/auth/login_company.html:
```diff
@@ -17,7 +17,7 @@ <h3 class="title">{{ gettext("Login company") }}</h3>
                             {{ form.email(class_="input is-large", placeholder=gettext("Email")) }}
                             <span class="icon is-small is-left">
-                                <i class="fas fa-at"></i>
+                                {{ "at"|icon }}
```
OK. Also, please note my answer above ☺️

Amittai-Aviram commented:
Thanks for your responses, Lexi!  About the "@" — in general, I really hate using images files (even SVGs) instead of plain text, but it sounds as if you have a good rationale.

Amittai-Aviram commented:
Sorry!  I just wanted to approve this PR and could not find an "approve" button.  I did not know what "close" meant.  This UI, like most Web UIs, is mostly stupid and opaque.

4lm commented:
> Sorry! I just wanted to approve this PR and could not find an "approve" button. I did not know what "close" meant. This UI, like most Web UIs, is mostly stupid and opaque.

No problem, no harm was done ☺️ UI done right, is a hard problem.

sloschert commented on arbeitszeit_flask/templates/macros/tooltips.html:
```diff
@@ -2,9 +2,9 @@
     <div class="dropdown-trigger">
         <span class="icon" aria-haspopup="true" aria-controls="{{ id }}">
-            <i class="fa-solid fa-info-circle">
+            {{ "info-circle"|icon }}
```
I think there's something wrong, isn't it? In chrome and firefox it shows like this:
![image](https://github.com/user-attachments/assets/ccceb50c-5ef8-4ae4-b170-b50c8f04a686)

4lm replied:
Thanks for noticing this! It went under in the super repetitive text search and replacement frenzy I had with all those icons 😉 I can directly spot the reason for this tooltip layout issue. I will fix it and come back to you.

sloschert commented on arbeitszeit_flask/templates/macros/transactions.html:
```diff
@@ -38,7 +38,7 @@
                 <strong class="is-size-5">
                     {{ type }}
-                </strong><span class="icon"><i class='{{ peer_type_icon }}'></i></span><span>{{ peer_name }}</span>
+                </strong><span class="icon">'{{ peer_type_icon|icon }}'</span><span>{{ peer_name }}</span>
```
This raises an error, when one tries to visit the "account prd" page of a company (e.g. http://127.0.0.1:5000/user/company/73880350-173b-472f-9c89-6278ff4f7293/account_prd):

```
raise FileNotFoundError(f'Icon "{icon_name}" not found: {str(e)}')
FileNotFoundError: Icon "" not found: [Errno 2] No such file or directory: 'arbeitszeit_flask/templates/icons/.html'
```

4lm replied:
Fixed, `icon_filter` now treats empty strings as intentionally set null values

sloschert commented on tests/flask_integration/test_filters.py:
```diff
@@ -0,0 +1,108 @@
+    def tearDown(self):
+        super().tearDown()
+
+    def test_icon_with_default_attributes(self):
```
In general we try to give our test functions more specific names, including for example a predicate/verb. I think the longer function names are justified by the better readability and by the fact that we do not call these test functions anywhere manually. It also forces us to think about what functionality we are testing exactly in each of these functions. Please let me know if you disagree.

4lm replied:
You are right 👍  I will improve the test case files. I also, will:

- remove unused `setUp` and `tearDown` functions
- double-check if more unit-tests are needed

Question. Are you OK with the decision that I mixed (Flask layer) unit tests of `icon_filter` with the integration test of the registered filter `icon` which is derived from `icon_filter`? Couldn't find a good place for the (Flask layer) unit tests in the `tests/*` folder.

sloschert replied:
> Are you OK with the decision that I mixed (Flask layer) unit tests of `icon_filter` with the integration test

I wouldn't know, where to put it elsewhere, so I'm fine with it.

sloschert reviewed (changes requested):
Really good work! I left a few comments. There is also a conflict that must be resolved.

4lm commented:
> Really good work! I left a few comments. There is also a conflict that must be resolved.

Thank you. I think I fixed all your points in question. Please note that I added my plan ID with all hours worked until this PR in the initial comment of this PR. The number is so high because this is my first PR with a plan ID, and it accounts for all work up until this PR (development of this PR, dev-meetings, and reviewing other PRs).

sloschert commented on tests/flask_integration/test_filters.py:
```diff
@@ -0,0 +1,167 @@
+
+
+class IconFilterIntegrationTest(ViewTestCase):
+    def test_that_icon_filter_return_the_in_the_login_form_present_key_icon_correctly(
```
There's something wrong with the function name.

4lm replied:
Yes, indeed! Something went wrong here. I had some very long test function names which I tried to shorten via a little help of a LLM. Something must have gone wrong in the copy paste process 😅

4lm replied:
Fixed.

sloschert reviewed (changes requested):
Thanks for the changes! Only few and small issues left from my side.

4lm commented on tests/flask_integration/test_filters.py:
```diff
@@ -0,0 +1,167 @@
+        super().setUp()
+        self.mock_file_reader = Mock()
+
+    def tearDown(self):
```
OK. Will remove!

4lm replied:
Fixed.

4lm commented on arbeitszeit_flask/filters.py:
```diff
@@ -0,0 +1,111 @@
+        default_attributes = {
+            "data-icon": icon_name,
+            "width": "24px",
+            "height": "20px",
```
To exactly match the default FontAwesome icon size of 24 px (width) by 20 px (height) we had before. My guess would be that they (FontAwesome), in some cases, visually try to compensate for human vision because the human brain sees equal x y length longer in the y direction. In psychology, this phenomenon is called [vertical–horizontal illusion](https://en.wikipedia.org/wiki/Vertical%E2%80%93horizontal_illusion).

4lm replied:
Left as is.

4lm commented on arbeitszeit_flask/templates/macros/plan_details.html:
```diff
@@ -129,7 +125,7 @@ <h1 class="title">
                         {% if view_model.details.price_per_unit[2] %} <span>
                             <a href="{{ view_model.details.price_per_unit[3] }}">
-                                <i class="fas fa-hands-helping"></i></a>
+                                {{ "hands-helping"|icon }}/a>
```
Will fix. Thanks for spotting.

4lm replied:
Fixed.

sloschert reviewed (approved):
Great, thanks! 29xCR
